Encode the runtime arch in all centos7 buildbox images

wadells opened the pull request:
Architecture suffixes were added in https://github.com/gravitational/teleport/pull/34950 and https://github.com/gravitational/teleport/pull/36111, but the change was missed in `images.mk`.  This resulted in suffix-less images when images.mk was used on its own, causing errors like the following:

``` Error: buildx failed with: ERROR: failed to solve: ghcr.io/gravitational/teleport-buildbox-centos7:teleport16: ghcr.io/gravitational/teleport-buildbox-centos7:teleport16: not found```

As seen at https://github.com/gravitational/teleport.e/actions/runs/7589597217/job/20674573771

This fixes a bug latent in v15 and v16.  I noticed it when running a dev build on v16, because v16 never had a `teleportXX` tag published. v15's bustedness wouldn't show up until the `teleport15-amd64` tag sufficiently diverges from `teleport15`

## Testing Done
Underway.

* build: https://drone.platform.teleport.sh/gravitational/teleport/32939
* promote:

There are other failures latent in master (arm64 :eyes:) so it is a bit difficult to sort out.

wadells committed:
Encode the runtime arch in all centos7 buildbox images

This was partially done before, but it was missed in images.mk, which
resulted in builder errors when images.mk data was used on its own.
This resulted in errors like the following:

  Error: buildx failed with: ERROR: failed to solve: ghcr.io/gravitational/teleport-buildbox-centos7:teleport16:
    ghcr.io/gravitational/teleport-buildbox-centos7:teleport16: not found

wadells commented:
Draft for now -- until I finish testing.

camscale commented on build.assets/arch.mk:
```diff
@@ -0,0 +1,11 @@
+# These variables are extracted from build.assets/Makefile so they can be imported
+# by other Makefiles
+ifeq ($(origin ARCH), undefined) # avoid use of "ARCH ?= $(...)" as the lazy loading will repeatedly re-run the same shell command
+	ARCH := $(shell go env GOARCH)
```
This will unconditionally run `go` every time `make` is invoked and spit out errors if `go` is not installed. This currently happens and has caused confusion, so we can try to fix that now. How about this:

```
ifeq ($(origin GOARCH), undefined)
        GOARCH := $(go env GOARCH 2>/dev/null)
endif

ARCH = $(or $(GOARCH),$(error go not installed to determine ARCH))
```

This will set `GOARCH` to whatever `go` has if the `go` command exists, otherwise it will be empty. Then `$(ARCH)` is set to that using a recursive assignment, so it does not get evaluated unless expanded. if `$(GOARCH)` is empty, we get an error instead.

I realise now that what I said about the existing errors and confusion refer to the main Makefile not this build.assets/Makefile. Ultimately we should unify these I think, but for now at least this should surface errors early if the go toolchain is not installed and we try to use `go env GOARCH`.

wadells replied:
The risk here is we're turning a soft failure into a hard failure -- which then requires me to go follow up and fix the consumers of this Makefile that don't have go available.  While that is laudable, it is a scope creep for into fixing something that "isn't broken" (builds will succeed), in a PR that is otherwise fixing something very broken (no tag builds will succeed).

I'll take a look at this, but if it ends up flushing out a lot of failures, I'm going to pass on it for this PR.

wadells replied:
Updated in https://github.com/gravitational/teleport/pull/36974/commits/fb5949add35ac7c6b27c0964d9c7be0b3a26ce67.  This has passed my local testing, but we won't know for sure it is safe until after I finish a tag build and promote.

wadells replied:
testing shows that making this a hard fail is going to break a bunch of stuff:

https://github.com/gravitational/teleport.e/actions/runs/7618315809/job/20749749082

As such, I'm going to skip it for this PR.

camscale committed:
build: Update buildbox definitions for spacelift runner

Add a var to `build.assets/images.mk` that defines the centos7 buildbox
without the host architecture suffix, and print that in the
`show-buildbox-centos7-target` used by the spacelift runner CI job.

The spacelift runner needs to know the buildbox to use to build `tbot`,
but since it does a `docker buildx` multiarch build, it needs to be able
to select different architecture buildboxes itself using `$TARGETARCH`
and not just use the one for the host architecture.

camscale commented:
I have tested this with some fixed for the spacelift runner to work properly: https://drone.platform.teleport.sh/gravitational/teleport/33767

I've pushed the changes for the spacelift runner (and remove the commit that updated the version). I am doing another test build now as I needed to resolve some conflicts with master, currently going at https://drone.platform.teleport.sh/gravitational/teleport/33963
